RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc

```diff
@@ -0,0 +1,611 @@
+
+  // Trackster - Trackster link finding
+  // step 2: tracksters EM -> HAD, at lastLayerEE
+
+  std::vector<std::vector<unsigned>> tsNearAtInt(tracksters.size());
+  findTrackstersInWindow(tsPropIntColl, tsHadPropIntTiles, del_ts_em_had_, tracksters.size(), tsNearAtInt, true);
+
+  // step 1: tracksters HAD -> HAD, at lastLayerEE
+
+  std::vector<std::vector<unsigned>> tsHadNearAtInt(tracksters.size());
+  findTrackstersInWindow(tsHadPropIntColl, tsHadPropIntTiles, del_ts_had_had_, tracksters.size(), tsHadNearAtInt, true);
+
+  dumpLinksFound(tsNearTk, "track -> tracksters at layer 1");
+  dumpLinksFound(tsNearTkAtInt, "track -> tracksters at lastLayerEE");
+  dumpLinksFound(tsNearAtInt, "EM -> HAD tracksters at lastLayerEE");
+  dumpLinksFound(tsHadNearAtInt, "HAD -> HAD tracksters at lastLayerEE");
+
+  // make final collections
+
+  std::vector<TICLCandidate> chargedCandidates;
+  std::vector<TICLCandidate> chargedHadronsFromTk;
+  std::vector<int> chargedMask(tracksters.size(), 0);
+  for (unsigned &i : candidateTrackIds) {
+    if (tsNearTk[i].empty() && tsNearTkAtInt[i].empty()) {  // nothing linked to track, make charged hadrons
+      TICLCandidate chargedHad;
+      const auto &tk = tracks[i];
+      chargedHad.setCharge(tk.charge());
+      chargedHad.setPdgId(211 * tk.charge());
+      chargedHad.setTrackPtr(edm::Ptr<reco::Track>(tkH, i));
+      float energy = std::sqrt(tk.p() * tk.p() + mpion2);
+      chargedHad.setRawEnergy(energy);
+      math::PtEtaPhiMLorentzVector p4Polar(tk.pt(), tk.eta(), tk.phi(), mpion);
+      chargedHad.setP4(p4Polar);
+      chargedHadronsFromTk.push_back(chargedHad);
+      continue;
+    }
+
+    TICLCandidate chargedCandidate;
+    double total_raw_energy = 0.;
+    auto energyCompatible = [&](const Trackster &ts, const reco::Track tk) -> bool {
+      // compatible if accumulated energy does not
+      // exceed track momentum by more than threshold
+      double threshold = std::min(0.2 * ts.raw_energy(), 10.0);
+
+      if (LinkingAlgoBase::algo_verbosity_ > LinkingAlgoBase::Advanced)
+        if (!(total_raw_energy + ts.raw_energy() < tk.p() + threshold))
+          LogDebug("LinkingAlgoByDirectionGeometric")
+              << "energy incompatible : track p " << tk.p() << " trackster energy " << ts.raw_energy() << "\n";
+
+      return (total_raw_energy + ts.raw_energy() < tk.p() + threshold);
+    };
+    auto timeCompatible = [&](const Trackster &ts, const reco::TrackRef tk) -> bool {
+      // compatible if trackster time is within 3sigma of
+      // track time; compatible if either: no time assigned
+      // to trackster or track time quality is below threshold
+      double maxDeltaT = 3.;
+      double tsT = ts.time();
+      double tsTErr = ts.timeError();
+      double tkT = tkTime[tk];
+      double tkTErr = tkTimeErr[tk];
+
+      if (tsT == -99. or tkTimeQual[tk] < timing_quality_threshold_)
```

Comment:
is -99.f defined as a magic number for an uninitialized trackster time? if so, it would be good if it was defined as a constexpr to prevent confusion in the future.